grudge/array_context.py

```diff
@@ -317,23 +326,39 @@ class PytestPyOpenCLArrayContextFactoryWithHostScalars(
 def get_reasonable_array_context_class(
         lazy: bool = True, distributed: bool = True
         ) -> Type[ArrayContext]:
+    """Returns a reasonable :class:`PyOpenCLArrayContext` currently
+    supported given the constraints of *lazy* and *distributed*."""
     if lazy:
+        if not _HAVE_SINGLE_GRID_WORK_BALANCING:
+            from warnings import warn
+            warn("No device-parallel actx available, execution will be slow."
+                 "Please make sure you have the right branches for loopy "
+                 "(https://github.com/kaushikcfd/loopy/tree/pytato-array-context-transforms) "  # noqa
+                 "and meshmode "
+                 "(https://github.com/kaushikcfd/meshmode/tree/pytato-array-context-transforms).")  # noqa
```

Comment:
That's not what I meant, and I don't think this makes sense. This will warn anytime lazy-distributed is run with meshmode `main`, which IMO is silly: that's a perfectly fine way to do business. The problem arises when someone has the right meshmode branch and a loopy branch that doesn't support it.

Reply:
> That's not what I meant, and I don't think this makes sense. This will warn anytime lazy-distributed is run with meshmode `main`, which IMO is silly: that's a perfectly fine way to do business.

But wouldn't that run single-threaded on a device, ie, really slow?

> The problem arises when someone has the right meshmode branch and a loopy branch that doesn't support it.

Ok, I'll add a warning for this case.

Reply:
> But wouldn't that run single-threaded on a device, ie, really slow?

If the device is a GPU, yes. For a CPUs and testing, it's not currently parallel but moderately sensible?

Reply:
> > The problem arises when someone has the right meshmode branch and a loopy branch that doesn't support it.
> 
> Ok, I'll add a warning for this case.

Added a warning in 41c5e98.

> > But wouldn't that run single-threaded on a device, ie, really slow?
> 
> If the device is a GPU, yes. For a CPUs and testing, it's not currently parallel but moderately sensible?

I'm not sure. I would guess this case (ie, both loopy and meshmode on their main branches) is not too uncommon, and since "everything works lazily+distributed", it is relatively easy to overlook that this will be quite slow for most use cases, even on CPU. The warning seems to me like a good compromise, and might incentivize us to merge those branches a bit quicker ;-).

Reply:
> might incentivize us to merge those branches a bit quicker ;-).

I see what you're doing there. :)